File: benchmarks/kdtree-raycast/lang_go/benchmark.go
package main

import (
	"common"
	"fmt"
	"math"
	"time"
)

const BenchmarkRaysCount = 10000000

func uniformSampleSphere() Vector64 {
	u1 := RandFloat64()
	u2 := RandFloat64()
	z := 1.0 - 2.0*u1
	r := math.Sqrt(1.0 - z*z)
	phi := 2.0 * math.Pi * u2
	x := r * math.Cos(phi)
	y := r * math.Sin(phi)
	return Vector64{x, y, z}
}

type rayGenerator struct {
	raysBounds BBox64
}

func newRayGenerator(meshBounds BBox64) *rayGenerator {
	diagonal := VSub64(meshBounds.maxPoint, meshBounds.minPoint)
	delta := 2.0 * VLength64(diagonal)

	raysBounds := NewBBox64FromPoints(
		VSub64(meshBounds.minPoint, NewVector64FromScalar(delta)),
		VAdd64(meshBounds.maxPoint, NewVector64FromScalar(delta)),
	)

	return &rayGenerator{
		raysBounds: raysBounds,
	}
}

func (rg *rayGenerator) generateRay(lastHit Vector64, lastHitEpsilon float64) Ray {
	// generate ray origin
	origin := Vector64{
		RandForRange(rg.raysBounds.minPoint[0], rg.raysBounds.maxPoint[0]),
		RandForRange(rg.raysBounds.minPoint[1], rg.raysBounds.maxPoint[1]),
		RandForRange(rg.raysBounds.minPoint[2], rg.raysBounds.maxPoint[2]),
	}

	useLastHit := RandFloat64() < 0.25
	if useLastHit {
		origin = lastHit
	}

	// generate ray direction
	direction := uniformSampleSphere()

	if RandFloat64() < 1.0/32.0 && direction[2] != 0.0 {
		direction[0] = 0
		direction[1] = 0
	} else if RandFloat64() < 1.0/32.0 && direction[1] != 0.0 {
		direction[0] = 0
		direction[2] = 0
	} else if RandFloat64() < 1.0/32.0 && direction[0] != 0.0 {
		direction[1] = 0
		direction[2] = 0
	}
	direction = VNormalized64(direction)

	// initialize ray
	ray := RayFromOriginAndDirection(origin, direction)

	if useLastHit {
		ray.Advance(lastHitEpsilon)
	} else {
		ray.Advance(1e-3)
	}
	return ray
}

func BenchmarkKdTree(kdTree *KdTree) int {
	start := time.Now()

	lastHit := VMul64(VAdd64(kdTree.meshBounds.minPoint, kdTree.meshBounds.maxPoint), 0.5)
	lastHitEpsilon := 0.0

	rg := newRayGenerator(kdTree.meshBounds)

	for raysTested := 0; raysTested < BenchmarkRaysCount; raysTested++ {
		ray := rg.generateRay(lastHit, lastHitEpsilon)

		hitFound, intersection := kdTree.Intersect(&ray)
		if hitFound {
			lastHit = ray.GetPoint(intersection.t)
			lastHitEpsilon = intersection.epsilon
		}

		// // debug output
		// if raysTested < 1024 {
		// 	if hitFound {
		// 		fmt.Printf("%d: found: %v, lastHit: %.14f %.14f %.14f\n",
		// 			raysTested, hitFound, lastHit[0], lastHit[1], lastHit[2])
		// 	} else {
		// 		fmt.Printf("%d: found: %v\n", raysTested, hitFound)
		// 	}
		// }
	}
	return int(time.Since(start) / time.Millisecond)
}

func ValidateKdTree(kdTree *KdTree, raysCount int) {
	lastHit := VMul64(VAdd64(kdTree.meshBounds.minPoint, kdTree.meshBounds.maxPoint), 0.5)
	lastHitEpsilon := 0.0

	rg := newRayGenerator(kdTree.meshBounds)

	for raysTested := 0; raysTested < raysCount; raysTested++ {
		ray := rg.generateRay(lastHit, lastHitEpsilon)

		kdTreeHitFound, kdTreeIntersection := kdTree.Intersect(&ray)

		bruteForceIntersection := KdTreeIntersection{t: math.Inf(+1)}
		bruteForceHitFound := false

		for i := int32(0); i < kdTree.mesh.GetTrianglesCount(); i++ {
			indices := kdTree.mesh.triangles[i]

			triangle := Triangle{[3]Vector64{
				NewVector64FromVector32(kdTree.mesh.vertices[indices[0]]),
				NewVector64FromVector32(kdTree.mesh.vertices[indices[1]]),
				NewVector64FromVector32(kdTree.mesh.vertices[indices[2]]),
			}}

			hitFound, intersection := IntersectTriangle(&ray, &triangle)

			if hitFound && intersection.t < bruteForceIntersection.t {
				bruteForceIntersection.t = intersection.t
				bruteForceHitFound = true
			}
		}

		if kdTreeHitFound != bruteForceHitFound ||
			kdTreeIntersection.t != bruteForceIntersection.t {
			o := ray.origin
			d := ray.direction
			fmt.Printf("KdTree accelerator test failure:\n"+
				"KdTree hit: %v\n"+
				"actual hit: %v\n"+
				"KdTree T %.16g [%b]\n"+
				"actual T %.16g [%b]\n"+
				"ray origin: (%b, %b, %b)\n"+
				"ray direction: (%b, %b, %b)\n",
				kdTreeHitFound, bruteForceHitFound,
				kdTreeIntersection.t, kdTreeIntersection.t,
				bruteForceIntersection.t, bruteForceIntersection.t,
				o[0], o[1], o[2], d[0], d[1], d[2])
			common.ValidationError("kdTree traversal error detected")
		}

		if bruteForceHitFound {
			lastHit = ray.GetPoint(bruteForceIntersection.t)
			lastHitEpsilon = bruteForceIntersection.epsilon
		}
	}
}

File: benchmarks/kdtree-raycast/lang_cpp/benchmark.cpp
#include "benchmark.h"
#include "common.h"
#include "kdtree.h"
#include "random.h"
#include "triangle.h"
#include "vector.h"
#include <cassert>

namespace {
const double PI = 3.14159265358979323846;

Vector UniformSampleSphere()
{
  auto u1 = RandDouble();
  auto u2 = RandDouble();
  assert(u1 >= 0.0 && u1 < 1.0);
  assert(u2 >= 0.0 && u2 < 1.0);

  double z = 1.0 - 2.0 * u1;
  double r = sqrt(1.0 - z * z);
  double phi = 2.0 * PI * u2;
  double x = r * cos(phi);
  double y = r * sin(phi);
  return Vector(x, y, z);
}

class RayGenerator {
public:
  RayGenerator(const BoundingBox& meshBounds)
  {
    auto diagonal = meshBounds.maxPoint - meshBounds.minPoint;
    double delta = 2.0 * diagonal.Length();

    raysBounds = BoundingBox(meshBounds.minPoint - Vector(delta),
                             meshBounds.maxPoint + Vector(delta));
  }

  Ray GenerateRay(const Vector& lastHit, double lastHitEpsilon) const
  {
    // generate ray origin
    Vector origin;
    origin.x = RandFromRange(raysBounds.minPoint.x, raysBounds.maxPoint.x);
    origin.y = RandFromRange(raysBounds.minPoint.y, raysBounds.maxPoint.y);
    origin.z = RandFromRange(raysBounds.minPoint.z, raysBounds.maxPoint.z);

    const bool useLastHit = RandDouble() < 0.25;
    if (useLastHit)
      origin = lastHit;

    // generate ray direction;
    auto direction = UniformSampleSphere();

    if (RandDouble() < 1.0 / 32.0 && direction.z != 0.0)
      direction.x = direction.y = 0.0;
    else if (RandDouble() < 1.0 / 32.0 && direction.y != 0.0)
      direction.x = direction.z = 0.0;
    else if (RandDouble() < 1.0 / 32.0 && direction.x != 0.0)
      direction.y = direction.z = 0.0;
    direction = direction.GetNormalized();

    // initialize ray
    auto ray = Ray(origin, direction);

    if (useLastHit) {
      ray.Advance(lastHitEpsilon);
    }
    else {
      ray.Advance(1e-3);
    }
    return ray;
  }

private:
  BoundingBox raysBounds;
};
} // namespace

int BenchmarkKdTree(const KdTree& kdTree)
{
  Timer timer;

  Vector lastHit =
    (kdTree.GetMeshBounds().minPoint + kdTree.GetMeshBounds().maxPoint) * 0.5;
  double lastHitEpsilon = 0.0;
  auto rayGenerator = RayGenerator(kdTree.GetMeshBounds());

  for (int raysTested = 0; raysTested < benchmarkRaysCount; raysTested++) {
    const Ray ray = rayGenerator.GenerateRay(lastHit, lastHitEpsilon);

    KdTree::Intersection intersection;
    bool hitFound = kdTree.Intersect(ray, intersection);

    if (hitFound) {
      lastHit = ray.GetPoint(intersection.t);
      lastHitEpsilon = intersection.epsilon;
    }

    //// debug output
    //if (raysTested < 1024) {
    //  if (hitFound)
    //    printf("%d: found: %s, lastHit: %.14f %.14f %.14f\n", raysTested,
    //           hitFound ? "true" : "false", lastHit.x, lastHit.y, lastHit.z);
    //  else
    //    printf("%d: found: %s\n", raysTested, hitFound ? "true" : "false");
    //}
  }
  return timer.ElapsedMilliseconds();
}

void ValidateKdTree(const KdTree& kdTree, int raysCount)
{
  Vector lastHit =
      (kdTree.GetMeshBounds().minPoint + kdTree.GetMeshBounds().maxPoint) * 0.5;
  double lastHitEpsilon = 0.0;
  auto rayGenerator = RayGenerator(kdTree.GetMeshBounds());

  for (int raysTested = 0; raysTested < raysCount; raysTested++) {
    const Ray ray = rayGenerator.GenerateRay(lastHit, lastHitEpsilon);

    KdTree::Intersection kdTreeIntersection;
    bool kdTreeHitFound = kdTree.Intersect(ray, kdTreeIntersection);

    KdTree::Intersection bruteForceIntersection;
    bool bruteForceHitFound = false;

    for (int32_t i = 0; i < kdTree.GetMesh().GetTrianglesCount(); i++) {
      const auto& p = kdTree.GetMesh().triangles[i].points;

      Triangle triangle = {
          {Vector(kdTree.GetMesh().vertices[p[0].vertexIndex]),
           Vector(kdTree.GetMesh().vertices[p[1].vertexIndex]),
           Vector(kdTree.GetMesh().vertices[p[2].vertexIndex])}};

      Triangle::Intersection intersection;
      bool hitFound = IntersectTriangle(ray, triangle, intersection);

      if (hitFound && intersection.t < bruteForceIntersection.t) {
        bruteForceIntersection.t = intersection.t;
        bruteForceHitFound = true;
      }
    }

    if (kdTreeHitFound != bruteForceHitFound ||
        kdTreeIntersection.t != bruteForceIntersection.t) {
      const auto& o = ray.GetOrigin();
      const auto& d = ray.GetDirection();
      printf("KdTree accelerator test failure:\n"
             "KdTree hit: %s\n"
             "actual hit: %s\n"
             "KdTree T %.16g [%a]\n"
             "actual T %.16g [%a]\n"
             "ray origin: (%a, %a, %a)\n"
             "ray direction: (%a, %a, %a)\n",
             kdTreeHitFound ? "true" : "false",
             bruteForceHitFound ? "true" : "false", kdTreeIntersection.t,
             kdTreeIntersection.t, bruteForceIntersection.t,
             bruteForceIntersection.t, o.x, o.y, o.z, d.x, d.y, d.z);
      ValidationError("KdTree traversal error detected");
    }

    if (bruteForceHitFound) {
      lastHit = ray.GetPoint(bruteForceIntersection.t);
      lastHitEpsilon = bruteForceIntersection.epsilon;
    }
  }
}
